woocommerce/woocommerce
# Performance: product sorting lookup table

This PR aims to improve performance of frontend sorting + filtering queries (e.g. price filter) by introducing a new lookup table which contains price, rating, and sales information for products.

Schema:

```php
CREATE TABLE {$wpdb->prefix}wc_product_meta_lookup (
  `product_id` bigint(20) NOT NULL,
  `min_price` decimal(10,{$price_decimals}) NULL default NULL,
  `max_price` decimal(10,{$price_decimals}) NULL default NULL,
  `average_rating` decimal(10,2) NULL default 0,
  `total_sales` bigint(20) NULL default 0,
  `stock` bigint(20) NULL default NULL,
  PRIMARY KEY  (`product_id`),
  KEY product_id_price (`product_id`,`max_price`,`min_price`),
  KEY product_id_average_rating (`product_id`,`average_rating`),
  KEY product_id_total_sales (`product_id`,`total_sales`)
  ) $collate;
```

During install this table is created, and an update routine moves meta data into this new table in bulk (this can take a while on larger stores but is ran once only). Only a single query is needed with joins.

When a product is updated via the CRUD, the data store handles updating this table also.

During frontend queries, this new table is joined to wp_posts by ID so it can be used for sorting and filtering. This replaces meta queries which were responsible for most of the slowdown. I've implemented this for sorting and the price filter widget.

Whilst working on this I found some unused code which I stripped out in the following SHAs:
- a0e05f972ef28c1f6fa64936f6d4753543846532
- 6d78218f97c54dfff20b0dd5c2d6a73dd7691dc9

There were also some legacy meta updates in data store read method which I removed in 2ebfe73 to avoid the need for more table syncs. These should not be needed any more since things like ratings are persistent, not transient based.

## Testing

- Deactivate WC
- Manually set DB version to 3.5.5
- Activate
- Run updater
- Check frontend sorting options work
- Check price filtering works
- Update product, check change is relfected in new table

## Benchmarks

On a test install with 25,000 products there are several queries which are notably slow on the frontend. These include:

- The main WP query, which is worse when sorting by price/sales/rating
- The price filter widget, specifically [`get_filtered_price()`](https://github.com/woocommerce/woocommerce/blob/a4acf5643a62dd511abea60c5ed557280d518f1f/includes/widgets/class-wc-widget-price-filter.php#L123) which looks for prices within the main query results.

Running master, here are some sample times of these queries (running `RESET QUERY CACHE;` between tests).

| Page      | Main query speed | `get_filtered_price` speed | Total
| -------- | ------------------ | -------------------------- | ----- |
| Shop - Default sorting | 0.0894 | 1.0735 | 1.1629
| Shop - Date | 0.0444 | 1.3324 | 1.3768
| Shop - Price high to low | 1.2996 | 1.0631 | 2.3627
| Shop - Price low to high | 1.2853 | 1.3670 | 2.6523
| Shop - Rating | 0.6824 | 0.9097 | 1.5921
| Shop - Sales | 1.3917 | 0.9436 | 2.3353

The slowest ones were the meta based sorts (sales, price).

Next I added the postmeta indexes we've been chatting about to see if there was any noticable impact.

| Page      | Main query speed | `get_filtered_price` speed | Total
| -------- | ------------------ | -------------------------- | ----- |
| Shop - Default sorting | 0.0912 | 0.4232 | 0.5144
| Shop - Date | 0.0303 | 0.3285 | 0.3588
| Shop - Price high to low | 1.0168 | 0.6426 | 1.6594
| Shop - Price low to high | 1.6083 | 0.5547  | 2.163
| Shop - Rating | 0.4251 | 0.4135 | 0.8386
| Shop - Sales | 0.6327 | 0.8529  | 1.4856

Some good improvements, more so on the filtered price query. 

Finally, I tested with this PR and no indexes.

| Page      | Main query speed | `get_filtered_price` speed | Total
| -------- | ------------------ | -------------------------- | ----- |
| Shop - Default sorting |  0.0968 | 0.0465 | 0.1433
| Shop - Date | 0.0273  | 0.0468 | 0.0741
| Shop - Price high to low | 0.1134 | 0.0500 | 0.1634
| Shop - Price low to high | 0.0731  | 0.0614 | 0.1345
| Shop - Rating | 0.1110 | 0.0488 | 0.1598
| Shop - Sales |  0.0961  | 0.0513 | 0.1474

Consistently faster 🚀

The price, high to low, for example went down from 2.3627 -> 0.1634. That is a 93% improvement.

Indexes alone went from 2.3627 -> 1.6594 which is only a 30% improvement.

🚀🚀🚀🚀🚀🚀🚀
